### PR TITLE
feat: add split prepare for trainer

### PR DESCRIPTION
**Pull request by bwanglzu**

### CODE REVIEW

File: jina/types/arrays/search_ops.py [hunk 1]
```diff
@@ -112,3 +113,21 @@ def shuffle(self, seed: Optional[int] = None) -> 'DocumentArray':
...
+        :return: a dict where Documents with the same value on `attribute` are grouped together, their orders are
+        preserved from the original :class:`DocumentArray`.
+        """
+        from .document import DocumentArray
+
+        rv = defaultdict(DocumentArray)
+        for doc in self:
+            value = doc.tags.get(attribute)
+            if not value:
+                continue
```
Comment (nan-wang) [inline jina/types/arrays/search_ops.py:131]: We need to add a warning message when the `attribute` does not exist in the `Document`. At least this skipping behavior should be claimed in the docstring.
```suggestion
                continue
```

**Review by nan-wang — CHANGES_REQUESTED** [review]

File: jina/types/arrays/search_ops.py (hunk 1 above)
Comment (hanxiao) [reply jina/types/arrays/search_ops.py:131]: please only label this behavior in docstring

**Review by numb3r3 — APPROVED** [review]
LGTM

**Review by JoanFM — COMMENTED** [review]
Shouldn't this function be generalized to a `groupby` with a predicate as input?

**Comment by bwanglzu** [comment]
@JoanFM can you name an example?

**Comment by JoanFM** [comment]
> @JoanFM can you name an example?

Would it be natural:

```python

da = DocumentArray(...)

groups = da.groupby(lambda x: x.tags['category'])
```

**Comment by bwanglzu** [comment]
this also looks good to me, how do you think cc @nan-wang @hanxiao

**Comment by hanxiao** [comment]
groupby has diff semantics,  https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.groupby.html

> A groupby operation involves some combination of splitting the object, applying a function, and combining the results.

I would just call this function `split`

File: jina/types/arrays/search_ops.py [hunk 2]
```diff
@@ -112,3 +113,24 @@ def shuffle(self, seed: Optional[int] = None) -> 'DocumentArray':
         from .document import DocumentArray
 
         return DocumentArray(self.sample(len(self), seed=seed))
+
+    def split(self, attribute: str) -> Dict[Any, 'DocumentArray']:
```
Comment (hanxiao) [inline jina/types/arrays/search_ops.py:117]: If it is only implemented with splitting by tags, then this arg should be called as `tag`

```suggestion
    def split(self, tag: str) -> Dict[Any, 'DocumentArray']:
```

also in docstring

**Review by hanxiao — APPROVED** [review]

**Pull request merged**